_posts/2024-10-25-topn.md

```diff
@@ -0,0 +1,466 @@
+We can call this a _top N by column analysis._
+It is particularly messy to try to do this analysis with ordinary SQL!
+You would need a subquery or window function for every single column...
+In DuckDB, simply:
+
+```sql
+FROM lineitem
+SELECT 
+    max(COLUMNS(*), 3) AS "top_3_\0";
+```
+
+|     top_3_l_orderkey     |    top_3_l_partkey    | ... |   top_3_l_shipmode    |                               top_3_l_comment                                |
+|--------------------------|-----------------------|-----|-----------------------|------------------------------------------------------------------------------|
+| [600000, 600000, 599975] | [20000, 20000, 20000] | ... | [TRUCK, TRUCK, TRUCK] | [zzle. slyly, zzle. quickly bold a, zzle. pinto beans boost slyly slyly fin] |
+
+## Top N by Group in DuckDB
+
+Armed with the new `N` parameter, how can we speed up a top N by group analysis?
+
+Want to cut to the chase and see the final output?
+[Feel free to skip ahead!](#the-final-top-n-by-group-query)
+
+We will take advantage of three other DuckDB SQL features to make this possible:
+
+* The [`max_by` function]({% link docs/sql/functions/aggregates.md %}#max_byarg-val-n) (also known as `arg_max`)
+* The [`unnest` function]({% link docs/sql/query_syntax/unnest.md %})
+* Automatically packing an entire row into a [`STRUCT` column]({% link docs/sql/data_types/struct.md %}#creating-structs)
+
+The `max` function will return the max (or now the max N!) of a specific column.
+In contrast, the `max_by` function will find the maximum value in a column, and then retrieve a value from the same row, but a different column.
+For example, this query will return the ids of the 3 most recently shipped orders for each supplier:
+
+```sql
+FROM lineitem
+SELECT 
+    l_suppkey,
+    max_by(l_orderkey, l_shipdate, 3) AS recent_orders
+GROUP BY
+    l_suppkey;
+```
+
+| l_suppkey |        recent_orders        |
+|----------:|-----------------------------|
+| 2992      | [233573, 3597639, 3060227]  |
+| 8516      | [4675968, 5431174, 4626530] |
+| 3205      | [3844610, 4396966, 3405255] |
+| 2152      | [1672000, 4209601, 3831138] |
+| 1880      | [4852999, 2863747, 1650084] |
+| ...       | ...                         |
+
+The `max_by` function is an aggregate function, so it takes advantage of DuckDB's fast hash aggregation rather than sorting.
+Instead of sorting by `l_shipdate`, the `max_by` function scans through the dataset just once and keeps track of the `N` highest `l_shipdate` values.
+It then returns the order id that corresponds with each of the most recent shipment dates.
+The radix sort in DuckDB must scan through the dataset once per byte, so scanning only once providing a significant speedup.
+For example, if sorting by a 64-bit integer, the sort algorithm must loop through the dataset 8 times vs. 1 with this approach!
+A simple benchmark is included below.
+
```

Comment:
There seems to be a missing link here in the text. Should a benchmark come here?

Reply:
Oh, the benchmark is just the results table and the included Python script. Definitely not a formal benchmark. Happy to reword - I'll take a look and see if I can rephrase.

Reply:
My only issue is that the "benchmark below" is three screens down. I think adding a link `[benchmark below](#performance-comparisons)` would resolve this.